src/firmware/host/types.rs

```diff
@@ -492,32 +505,50 @@ pub struct SnpExtConfig {
     /// GET:
     ///     Address of extended guest request certificate chain or None when
     ///     certificate should not be fetched.
-    pub certs: Option<CertTable>,
+    pub certs: Option<Vec<CertTableEntry>>,
 
     /// SET:
     ///     Length of the certificates.
     ///
     /// GET:
     ///     Length of the buffer which will hold the fetched certificates.
-    pub certs_len: u32,
+    pub certs_buf: u32,
```

Comment:
My issue with using `certs_len`, is that it holds two meanings. First, the length of the buffer being pointer to by the `certs_address` field (the correct one). Second, the number of certificates being supplied. I am open to keeping it as certs_len, though.

Reply:
The [documentation for the SNP_GET_EXT_REPORT ioctl](https://github.com/torvalds/linux/blob/master/Documentation/virt/coco/sev-guest.rst#23-snp_get_ext_report) says:

> On success, the snp_ext_report_resp.data will contain the attestation report and snp_ext_report_req.certs_address will contain the certificate blob. If the length of the blob is smaller than expected then snp_ext_report_req.certs_len will be updated with the expected value.

So I think that the `certs_len` field always holds length in bytes (never number of certs). When calling it holds the number of bytes in the certs_addresses buffer, and if that's not enough, on output it will hold how many bytes are needed.

Similarly for `SNP_GET_EXT_CONFIG` (hypervisor command).

So I'd go with `certs_len` -- which also matches the names exposed in the [kernel headers](https://github.com/torvalds/linux/blob/7287904c8771b77b9504f53623bb477065c19a58/include/uapi/linux/sev-guest.h#L59-L67):

```c
struct snp_ext_report_req {
	struct snp_report_req data;

	/* where to copy the certificate blob */
	__u64 certs_address;

	/* length of the certificate blob */
	__u32 certs_len;
};
```

Reply:
Why include it at all?

A suggestion: For the user API, get rid of this value altogether. When the API is issued, the library can traverse each cert in the `Vec<CertTableEntry>`, get the size of the cert, and add all of these sizes up (along with rounding up to the nearest page size). This sum will be the value in `certs_len` for the `SnpExtReport` request.

Reply:
@tylerfanelli It isn't in the user api, it is parsed from the user api into the kernel specific structure. 

@dubek Switched back to `certs_len` in next commit.